### ISSUE TITLE
AndroidTV - DASH/Widevine - Unable to use 2 video decoders simultaneously on some recent devices - "MediaCodec$CryptoException: Operation not supported in this configuration"

### ISSUE THREAD

**Issue by kvillnv** [google/ExoPlayer#9707]
Hi,

There is a new issue spreading on some devices since this summer:
* Sony Bravia KD-75X85J AndroidTV 10 `Build.DEVICE = BRAVIA_VH2`
* Sony Bravia XBR-65X900H AndroidTV 10 `Build.DEVICE = BRAVIA_VH1`
* Hisense 65U8G AndroidTV 10 (detected this week)

I am using the latest exoplayer version: 2.16.0
While looking for regressions in my project, I went all the way back to exoplayer 2.9.6 and still had the bug.

I also found a workaround that I will explain.

---------

### Search existing issues, including issues that are closed:
(Issues with the same error but different problem)
- https://github.com/google/ExoPlayer/issues/4487 
- https://github.com/google/ExoPlayer/issues/4101 
- https://github.com/google/ExoPlayer/issues/8037

### Clear reproduction steps including observed and expected behavior
1. The main player is playing a DASH H264 Widevine stream in 720p or 1080p on a `SurfaceView`. This player uses the first secure hardware decoder of the device.
2. On top of the main player We start a PIP, DASH H264 Widevine 360p on a `TextureView`, audio disabled. This player uses the first non-secure decoder available.
3. **Actual result:** The PIP starts and plays, but the main player stops playing and throws a `MediaCodecVideoRenderer error`caused by: `android.media.MediaCodec$CryptoException: Operation not supported in this configuration` (more details in logcat files)
**Expected result:** Same as all other devices: Show the PIP and keep playing the main player.

```
// Error in the main player (which was playing correctly before starting the PIP)

11-11 21:02:18.205 10043 10959 D PlayerCodecSelector: video/avc -> [OMX.MTK.VIDEO.DECODER.AVC.secure]
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal: Playback error
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:   com.google.android.exoplayer2.ExoPlaybackException: MediaCodecVideoRenderer error, index=0, format=Format(2, null, null, video/avc, avc1.4D4020, 2499968, null, [960, 540, 59.94006], [-1, -1]), format_supported=YES
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:558)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at android.os.Handler.dispatchMessage(Handler.java:103)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at android.os.Looper.loop(Looper.java:214)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at android.os.HandlerThread.run(HandlerThread.java:67)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:   Caused by: android.media.MediaCodec$CryptoException: Operation not supported in this configuration
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at android.media.MediaCodec.native_queueSecureInputBuffer(Native Method)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at android.media.MediaCodec.queueSecureInputBuffer(MediaCodec.java:2706)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.mediacodec.SynchronousMediaCodecAdapter.queueSecureInputBuffer(SynchronousMediaCodecAdapter.java:148)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.feedInputBuffer(MediaCodecRenderer.java:1367)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:825)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:978)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:482)
11-11 21:02:18.209 10043 10959 E ExoPlayerImplInternal:       ... 3 more
```

I am sending adb logcats with a lot more information at dev.exoplayer@gmail.com

* It is surprising to see the main player failing instead of the PIP which is the one trying to initialize. Having the error in the PIP would have been better for handling.
* During my troubleshooting on `BRAVIA_VH1`  I also tried to force the second non-secure decoder `c2.android.avc.decoder` instead of the default `OMX.MTK.VIDEO.DECODER.AVC`. But still got the same outcome.
* So far, since August we received complains only on the 3 devices I mentionned, the Hisense was reported this week, before that I thought it was Bravia specific (not all Bravias). Those devices all have great specs. They all run Android 10, but I doubt it is the reason.
* I can confirm the issue does not happen on all the other devices I tested: `Broadcom AOSP STB, Google ADT-3 (Android 11), Google TV chromecast, Many Nvidia Shields, FireTV 4K, FireTV Gen 2, many Bravia (not VH1 and VH2)`.
* It seems related to the Widevine CDM or Android revision. FireTV 4K uses same decoders than BRAVIA_VH1 and does not have the bug. Decoder: `OMX.MTK.VIDEO.DECODER.AVC.secure`
* I noticed on most other Bravias the second software decoder is `OMX.google.h264.decoder` but VH1 and VH2 have `c2.android.avc.decoder` instead. (I could not verify on the customer's HiSense if it is the same)

### Workaround
The workaround was to force Widevine L3 on the PIP. 
I would prefer to avoid it, it creates bugs on other devices. I am explaining it here only because I think it can help troubleshooting.    
`Operation not supported in this configuration` seems related to the Device unable to play `widevine L1`, but I know for sure `L1` works in the main player and we are not expecting to use `L1` in the PIP. So I though maybe the system has a bug where it is not able to select `L3` for the PreviewPlayer (as expected)..

Initially, the MainPlayer and PreviewPlayer (PIP) were each using a simple DrmSessionManager (1 instance each):
`DefaultDrmSessionManager.Builder().build(delegatedDrmCallback)`
- I kept it for the MainPlayer
- And for the PreviewPlayer, inside the `DefaultDrmSessionManger.Builder`  I forced Widevine `L3` with in the `securityLevel` property.
- Now we get the expected behavior: both players can play simultaneously. I got confirmation on `BRAVIA_VH1` and `BRAVIA_VH2`.
But unfortunately on most of all the other devices it creates a ~1 second video/audio freeze in the MainPlayer while initializing the PIP... 
So the workaround has to be conditionnal to a specific list of devices. Not ideal... it will not handle unknown and new devices. 

There is a probability that more devices will be affected with upcoming OS updates.

---------------

If I missed something, if you have a fix or better workaround please let me know.
Even if it is a manufacturer/exoplayer/widevine issue I have to come up with a solution/workaround for my product. Any help would be appreciated. 
Thanks

**Comment by marcbaechinger** [comment]
Thanks for the detailed report. 
@rrfrias may I ask to look into this?

**Comment by icbaker** [comment]
@kvillnv Can you try an alternative workaround and see what happens: When building your `DefaultDrmSessionManager` instance for the PiP player, can you try calling [`Builder#setUseDrmSessionsForClearContent(C.TRACK_TYPE_AUDIO, C.TRACK_TYPE_VIDEO)`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/drm/DefaultDrmSessionManager.Builder.html#setUseDrmSessionsForClearContent(@com.google.android.exoplayer2.C.TrackTypeint...))?

That should cause ExoPlayer to use a secure decoder for playing the clear content in the PiP. I don't know if that will alleviate the ~1 second pause you see on other devices with your existing L3 workaround.

More info: https://exoplayer.dev/drm.html#drm-sessions-for-clear-content

**Comment by kvillnv** [comment]
@icbaker thanks for the suggestion.
* Applying both workarounds does still cause a audio/video freeze on other devices.
* Applying only your workaround does not cause a freeze on other devices. But I don't know yet the result on the faulty devices, probably monday.

If it does not work, my plan for now is to put my workaround in a switch in the video settings of the app.

**Comment by linhai326** [comment]
we had the same issue: as far as we know, it depends on the device hardware/firmware that how many hardware decoder instances it can support at the same time. For those devices which can support only one instance, we use software decoder from google, which is available on most if not all android devices, for PIP playback.

**Comment by kvillnv** [comment]
> we had the same issue: as far as we know, it depends on the device hardware/firmware that how many hardware decoder instances it can support at the same time. For those devices which can support only one instance, we use software decoder from google, which is available on most if not all android devices, for PIP playback.

I noticed the 2 faulty devices I could test (BRAVIA_VH1 and BRAVIA_VH2) do not have the standard `OMX.google.h264.decoder`, they have `c2.android.avc.decoder` instead. I already tried it but I got the same result.

**Comment by kvillnv** [comment]
@icbaker I got the result: Unfortunately `Builder#setUseDrmSessionsForClearContent(C.TRACK_TYPE_AUDIO, C.TRACK_TYPE_VIDEO)` did not work on BRAVIA_VH1. We still got the same error.

**Comment by icbaker** [comment]
FYI `OMX.google.h264.decoder` is an alias for `c2.android.avc.decoder`, so it's not surprising they're showing the same behaviour. From `adb shell dumpsys media.player` on a Pixel 5:
```
  Encoder "c2.android.avc.encoder" supports
    aliases: [
      "OMX.google.h264.encoder" ]
```

**Comment by kvillnv** [comment]
FYI: I did not apply the workaround. I just improved my error message to let users know their device does not support the current configuration and suggest to disable PIP and video preview in the app settings.
It's unfortunate to have less features on high-end devices.

Do you think it has to be solved on the manufacturer side, or is there a chance for a exoplayer fix/workaround?

**Comment by rrfrias** [comment]
Sorry, i lost track of this. Might it be possible to share an apk that reproduces the problem. We would like your permission to share the logcat you sent to dev.exoplayer@gmail.com with Sony. Let us know if you have any objections.

**Comment by kvillnv** [comment]
Yes you can share with Sony. 
Were you able to identify the cause?

Since November, we made the PIP an "Experimental feature" that people can enable while being warned it is not compatible with all devices, so it does not look like a bug anymore.

Providing an APK right now is not possible. I can add it in backlog. We would have to write an app for this case, or we'll have to modify our app.
An APK with a main player using hardware decoder + a secondary smaller player with software decoder on top should do the trick on a Bravia VH1. The main player needs to be already playing when starting the secondary player.

**Comment by rrfrias** [comment]
We are passing it to Sony for furthur investigation. 

Internal ref: 223469195

**Comment by rrfrias** [comment]
@kvillnv it will be helpful for Sony's investigation if you can share an apk that repros the issue.
